### ISSUE TITLE
[BUG] Disks become over provisioned when Storage Over Provisioning Percentage is set to 100

### ISSUE THREAD

**Issue by hoo29** [longhorn/longhorn#8450]
## Describe the bug

With Harvester 1.3.0 and Longhorn 1.6.0, we have observed several disks becoming over provisioned despite Storage Over Provisioning Percentage being set to 100.

## To Reproduce

Theoretical steps (we haven't reproduced)

1. Provision enough VMs in harvester to nearly saturate all available disk space.
2. Delete the VMs but don't delete their volumes (in terraform have auto_delete set to false)
3. Create all the VMs again.

## Expected behavior

Disks do not become over provisioned and VMs fail to schedule if there isn't enough storage.

## Support bundle for troubleshooting

Please can no URLs or VM details from the bundle be posted to this issue.

Sent to longhorn-support-bundle@suse.com. 

## Environment

 - Longhorn version: 1.6.0
 - Impacted volume (PV): 
   harvester-node-2 disk /var/lib/harvester/extra-disks/0e75b3ff4813c3cae0f71a1e9f3ac893
   harvester-node-5 disk /var/lib/harvester/extra-disks/89ade731face5f52e750ade464ca09bc
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Harvester 1.3.0 ISO install
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: The RKE2 bundled with Harvester 1.3.0
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 5
 - Node config
   - OS type and version: SLE Micro 5.4
   - Kernel version:  5.14.21-150400.24.108-default
   - CPU per node: Xeon Gold 5320T, 20 CPU Cores
   - Memory per node: 384GB
   - Disk type (e.g. SSD/NVMe/HDD): NVMe SSD
   - Network bandwidth between the nodes (Gbps): 1 Gbps (test environment)
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal
 - Number of Longhorn volumes in the cluster: 83 - After clean up of dangling volumes

**Comment by james-munson** [comment]
Looking.  The first thing I notice is that in the Harvester settings, there is an `overcommit-config` with a storage value of 200:
```yaml
- apiVersion: harvesterhci.io/v1beta1
  default: '{"cpu":1600,"memory":150,"storage":200}'
  kind: Setting
  metadata:
...
   name: overcommit-config
    resourceVersion: "13529643"
    uid: fa02c2bd-f32b-4872-b803-112aec13351d
  status: {}
```

**Comment by ejweber** [comment]
I discussed this with @james-munson offline and took a look at the support bundle. The bundle can't be posted here, but using the various generic object names:

For one of the overscheduled disks `cbe58e05af80439c9320336f1dbb5dfc`, there are 12 replicas scheduled. The `CreationTimestamp` for each replica is somewhat worthless for debugging, since most of them were created as clones of their original during a migration. However, we can see that 11 out of the 12 volumes with scheduled replicas were created during the exact same second (`2024-04-17T15:40:44Z`):

- pvc-723ccacf-9491-4e7f-8809-5eee63cb0216
- pvc-49617dbb-6077-4428-80e6-24000231a0a8
- pvc-4e19d165-72c8-43d7-8570-7607ee206b23
- pvc-16fdd3fd-cc3b-496b-b995-1f4458ac1d98
- pvc-39b8415a-c91a-4a5b-872a-325058fb6815
- pvc-99700ed4-00c4-4154-b9aa-c4c756649ad6
- pvc-9943036e-8427-4b48-bb26-e788acfef2e1
- pvc-b5d7b3c3-a57b-493b-a33d-d4ce528a6789
- pvc-b4040b28-58fe-4b0c-9e43-b73dfe74c050
- pvc-cba8733b-573a-4355-8e9c-4efae44116b2
- pvc-d07ae34f-7bb7-42f8-ae59-dbc2d893b65e

@PhanLe1010 and I (and probably others) discussed this while I was working on https://github.com/longhorn/longhorn/issues/8043, but we need a followup ticket for it. That issue was more specific, but in general, we think Longhorn is vulnerable to accidental overscheduling if it is scheduling replicas for multiple volumes simultaneously.

The general flow of the replica scheduling is:

- The volume controller that owns a volume decides to schedule a replica.
- The volume controller looks at all the nodes to see where the replica can be scheduled.
- The volume controller decides on a node to schedule the replica to.
- The volume controller updates the replica with its decision.
- The node controller that owns the node later looks at the replicas and realizes that a new replica has been scheduled.
- The node controller updates the node to reflect it.
- The next volume controller that schedules a replica sees the updated node information.

Now, imagine two different volume controllers are scheduling for two different volumes at the same time.

- Both volume controllers see the same nodes.
- Both volume controllers think a particular disk on a particular node is a good scheduling choice.
- Both volume controllers decide to schedule a replica to that disk.
- Both volume controller update their own replica. There is not conflict, because they are scheduling different replicas for different volumes.
- Later, the node controller updates the node to reflect both replicas were scheduled.
- The node controller reports an overscheduling, but it is just informational.

In summary, I believe this happened as a result of many volumes being created simultaneously. We need to improve Longhorn replica scheduling to ensure it cannot happen.

**Comment by ejweber** [comment]
### Workaround to avoid the issue

This issue seems to be quite rare. It is probably because:

- Many volumes are rarely created simultaneously.
- Even when they are, not all replicas will be scheduled to the same disk.
- Even when they are, the replicas may fit anyway.

I am not sure why all the volumes were created simultaneously in this ticket. Perhaps there was some other factor at work.

If your workflow involves creating many volumes simultaneously, it may be best to try intentionally slow it down a bit until a fix is implemented (e.g. create one volume, wait a second, create another volume).

### Workaround if you have hit the issue

It should be possible to evict individual replicas from the overscheduled disks. Longhorn will find a different disk and move the data. This can be done while the workload is running.

**Comment by james-munson** [comment]
The ticket for a general solution to pick a leader from all longhorn-managers to avoid races and conflicts is #5571

**Comment by hoo29** [comment]
Thank you for the detailed response and workaround. We are creating all of our machines and disks in one go with the harvester terraform provider; we'll add a pause somewhere to help avoid this issue.

Marking the disk as un-schedulable and using evicting allowed us to rebalance things.

I'm not clear on how the harvester overcommit settings factor into this. In Slack Connor Kuehl said
> Connor Kuehl :rancher_employee:  [2 hours ago](https://rancher-users.slack.com/archives/CC2UQM49Y/p1714150396867289?thread_ts=1714057595.884409&cid=CC2UQM49Y)
Yes, Harvester unconditionally overwrites Longhorn's overcommit with the value from Harvester's overcommit settings (storage)


Is it the case our disks became to over provisioned due to the harvester overcommit settings, the bug you have described, or both?

**Comment by ejweber** [comment]
This is a good question. My current belief is that:

- Harvester is incorrectly ignoring its own over-commit setting and not propagating it to Longhorn.
- Separately, Longhorn overscheduled the disks as a result of the bug I described.

This is because the support bundle clearly shows the the `over-commit` setting with its default of `200` and the Longhorn `storage-over-provisioning-percentage` setting with a value of `100`.

I will ask Connor to take a look at the first part and help decide if it is a Longhorn bug.

**Comment by james-munson** [comment]
This bug allows Longhorn to over-schedule some disks, where it clearly might have been possible to consider other disks.  If the Harvester value of 200% were used, Longhorn still might still have scheduled the same way, but not have reported them as over-scheduled later on.  Whether they actually are over-committed depends on how much data is written to the volumes over time.

I think we do still have work to do to ensure that the values are in step.

**Comment by innobead** [comment]
cc @derekbit

**Comment by hoo29** [comment]
We are consistently hitting this bug every time we try add a new VM, resize a disk, or do a migration. Is there anything you can suggest that will help us work around the problem?

Even if we evict disks Longhorn will frequently auto balance replicas back to the disk that was just evicted and over provision it.

**Comment by ejweber** [comment]
> Even if we evict disks Longhorn will frequently auto balance replicas back to the disk that was just evicted and over provision it.

Can you help clarify this @hoo29?

Generally Longhorn will only auto-balance replicas if there is currently a lack of redundancy. For example, if there are three replicas for a volume spread across only two nodes, and a third node becomes available. If, on the other hand, there are three replicas for a volume spread across three nodes and a new node becomes available, Longhorn will do nothing.

Is there a lack of redundancy before the evicted disk becomes available? And is the only valid scheduling location the overprovisioned node?

**Comment by hoo29** [comment]
@ejweber - Sorry saying replicas was likely misleading. To take one VM as an example, we have 6 x 500 GB volumes on a 1 replica storage class attached to it. We did have 3 x 1TB disks but changed it fewer smaller disks to see if it helped with the scheduling issues. The data locality is set to best effort.

A 1.8TB disk on the host the VM was scheduled on became over provisioned and had 5 of the 6 volumes scheduled on it. We evicted it until it was empty but as soon as schedulable was set back to true the 5 x 1 replica volumes got moved immediately back to the just evicted disk (I assume because of the data locality?).

Incidentally for testing, I built and deployed a custom build of longhorn-manager which reverted the change in https://github.com/longhorn/longhorn-manager/pull/1397 and made the scheduling use a random suitable disk on the host. It worked quite well but not something we want to deploy in production.

**Comment by ejweber** [comment]
@hoo29, can you provide a support bundle that demonstrates this behavior? For example, captured after you have set the node back to schedulable and the replicas have moved back?

Data locality is definitely a potential reason for this behavior. If you have it set to best-effort, we will always try to move a replica to the node that a volume's engine is running on. Perhaps this is working similarly to https://github.com/longhorn/longhorn/issues/8450#issuecomment-2079894403. We decide to move all best-effort replicas simultaneously and accidentally overprovision the node.

> Incidentally for testing, I built and deployed a custom build of longhorn-manager which reverted the change in https://github.com/longhorn/longhorn-manager/pull/1397 and made the scheduling use a random suitable disk on the host. It worked quite well but not something we want to deploy in production.

I see. So it's something like the following?

- Longhorn wants to move all the replicas owned by a VM back to the node the VM is running on simultaneously.
- One of the disks on the node has the most available space, so it is the best candidate for scheduling according to the above change.
- All replicas are scheduled to the disk, accidentally ignoring provisioning limits.

It sounds like the situation may be improved for now by disabling best-effort data locality. Unfortunately, a VM created with six volumes may still trigger the issue, but the replicas can be moved without Longhorn automatically moving them back.

**Comment by hoo29** [comment]
@ejweber  - I have sent the harvester support bundle to longhorn-support-bundle@suse.com. I assume this includes the longhorn pod logs but also have the longhorn bundle if you need it. Please can no URLs or VM details from the bundle be posted to this issue.

1. I evicted 3 disks from node 1 to try to recreate the problem.
2. Disk 4b749db81b6e510a665e05bb9b4617f7 on node 0 became over provisioned (not the specific issue I was trying to recreate)
3. I re-enabled scheduling on dfbf26813939563d3b2aad28eb337c5f on node 1 and it became over provisioned with the 4 * 500GB disks.
4. I evicted dfbf26813939563d3b2aad28eb337c5f and waited until longhorn showed storage scheduled as 0.
5. I re-enabled scheduling for dfbf26813939563d3b2aad28eb337c5f and it immediately became over provisioned with 4 * 500GB disks on it.

**Comment by ejweber** [comment]
Thanks @hoo29. Unfortunately, the `longhorn-manager-sw49r` logs from node 1 appear to have rotated out between the events described and the capture of the support bundle.  This is the longhorn-manager we would expect to be logging about scheduling issues on node 1.

However, it's reasonable to assume that https://github.com/longhorn/longhorn/issues/8450#issuecomment-2079894403 applies here.

From the bundle,:

- dfbf26813939563d3b2aad28eb337c5f is one of five disks that all are the same size on node 1. (The other two disks are smaller.)
- There is a VM running on node 1 that uses six `data-locality = best-effort` volumes that are the same size. 
- Four of these volumes together are too large for a single disk.
- At the moment the evicted disk becomes available, it has the most `storageAvailable` (since it has been completely evicted).
- As soon as the evicted disk becomes available, Longhorn tries to schedule replicas for all volumes that don't have a local replica on node 1 to the disk with the most `storageAvailable` on node 1.

**Comment by hoo29** [comment]
I am surprised more people aren't experiencing this given how often we hit it. I see this issue has been moved from the 1.7 milestone to 1.8; we will eagerly await that version for the fix!

**Comment by hoo29** [comment]
It looks like longhorn 1.7 & https://github.com/longhorn/longhorn/issues/4105 might help out with this a bit

**Comment by derekbit** [comment]
@ChanYiLin Please check again. If the issue cannot be reproduced, please close the ticket.

**Comment by hoo29** [comment]
@ChanYiLin  - did you manage to reproduce this issue?

**Comment by ChanYiLin** [comment]
Hi @hoo29 
I am picking up the issue now and have read through the context.

I think @ejweber 's comment makes sense.
The VM had 6 volumes (6 x 500GB), each with a single replica and a best-effort setting.
The host running the VM had only 1.8TB of available storage. Due to multiple Longhorn controllers attempting to move replicas of all the volumes to the same node simultaneously, a race condition occurred and the disk was over provisioned.

> Thanks [@hoo29](https://github.com/hoo29). Unfortunately, the `longhorn-manager-sw49r` logs from node 1 appear to have rotated out between the events described and the capture of the support bundle. This is the longhorn-manager we would expect to be logging about scheduling issues on node 1.
> 
> However, it's reasonable to assume that [#8450 (comment)](https://github.com/longhorn/longhorn/issues/8450#issuecomment-2079894403) applies here.
> 
> From the bundle,:
> 
> * dfbf26813939563d3b2aad28eb337c5f is one of five disks that all are the same size on node 1. (The other two disks are smaller.)
> * There is a VM running on node 1 that uses six `data-locality = best-effort` volumes that are the same size.
> * Four of these volumes together are too large for a single disk.
> * At the moment the evicted disk becomes available, it has the most `storageAvailable` (since it has been completely evicted).
> * As soon as the evicted disk becomes available, Longhorn tries to schedule replicas for all volumes that don't have a local replica on node 1 to the disk with the most `storageAvailable` on node 1.

I will try to find a way to consistently reproduce the issue.
The improvement from Longhorn side for this might be
1. to have a lock for cluster resources when doing the scheduling 
2. only allow one controller to do the scheduling (leader-election)
3. only schedule one replica at a time globally.

cc @derekbit 

You can also disable the best-effort setting for the volumes if you don't have a single disk large enough to store all the replicas. Alternatively, you could prioritize the volumes and apply the best-effort setting only to the most important volume.

**Comment by ChanYiLin** [comment]
I can reproduce a simple version of this issue

0. Set the over provisioned in the setting to 100
1. Disable all the disks in the cluster except for one, lets call it diskX
2. Prepare two PVCs which the sum of the size is large than the diskX (assum diskX is less than 60GB)
3. Two PVC needs to be in separate files. 

```
sc.yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: longhorn-test
provisioner: driver.longhorn.io
allowVolumeExpansion: true
reclaimPolicy: Delete
volumeBindingMode: Immediate
parameters:
  numberOfReplicas: "1"
  staleReplicaTimeout: "2880"
  fsType: "ext4"
  dataEngine: "v1"
```
```
# pvc1.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: volume1
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: longhorn-test
  resources:
    requests:
      storage: 30Gi
```
```
# pvc2.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: volume2
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: longhorn-test
  resources:
    requests:
      storage: 30Gi
```
4. Use two terminal to apply these two yamls at the same time.
5. There is a chance that both volume's replica are scheduled on the same disk and the disk was over provisioned.

Normally, if we apply two PVC separately, the second one will fail at pre-check and have the message
```
Replica Scheduling Failure
Error Message: disks are unavailable;insufficient storage;precheck new replica failed
```

**Comment by PhanLe1010** [comment]
> Is the intended fix for this going to be https://github.com/longhorn/longhorn/issues/10422 ?

Hi @hoo29 I think yes!
